File: lib/data/image.cc
#include "lib/core/globals.h"
#include "lib/data/sector.h"
#include "lib/data/image.h"
#include "lib/data/layout.h"
#include "lib/config/config.h"

Image::Image() {}

Image::Image(const std::vector<std::shared_ptr<const Sector>>& sectors)
{
    for (auto& sector : sectors)
        _sectors[{sector->logicalCylinder,
            sector->logicalHead,
            sector->logicalSector}] = sector;

    calculateSize();
}

void Image::clear()
{
    _sectors.clear();
    _geometry = {0, 0, 0};
}

void Image::createBlankImage()
{
    clear();
    for (const auto& trackAndHead : Layout::getTrackOrdering(
             globalConfig()->layout().filesystem_track_order()))
    {
        unsigned track = trackAndHead.first;
        unsigned side = trackAndHead.second;
        auto trackLayout = Layout::getLayoutOfTrack(track, side);
        Bytes blank(trackLayout->sectorSize);
        for (unsigned sectorId : trackLayout->naturalSectorOrder)
            put(track, side, sectorId)->data = blank;
    }
}

bool Image::empty() const
{
    return _sectors.empty();
}

bool Image::contains(const LogicalLocation& location) const
{
    return _sectors.find(location) != _sectors.end();
}

std::shared_ptr<const Sector> Image::get(const LogicalLocation& location) const
{
    auto i = _sectors.find(location);
    if (i == _sectors.end())
        return nullptr;
    return i->second;
}

std::shared_ptr<Sector> Image::put(const LogicalLocation& location)
{
    auto sector = std::make_shared<Sector>(location);
    _sectors[location] = sector;
    return sector;
}

void Image::erase(const LogicalLocation& location)
{
    _sectors.erase(location);
}

void Image::addMissingSectors(const DiskLayout& diskLayout)
{
    for (auto& location : diskLayout.logicalSectorLocationsInFilesystemOrder)
        if (!_sectors.contains(location))
        {
            auto& ltl = diskLayout.layoutByLogicalLocation.at(
                {location.logicalCylinder, location.logicalHead});
            auto sector = std::make_shared<Sector>(location);
            sector->status = Sector::MISSING;
            _sectors[location] = sector;
        }
    calculateSize();
}

void Image::calculateSize()
{
    _geometry = {};
    unsigned maxSector = 0;
    for (const auto& i : _sectors)
    {
        const auto& sector = i.second;
        if (sector)
        {
            _geometry.numCylinders = std::max(
                _geometry.numCylinders, (unsigned)sector->logicalCylinder + 1);
            _geometry.numHeads =
                std::max(_geometry.numHeads, (unsigned)sector->logicalHead + 1);
            _geometry.firstSector = std::min(
                _geometry.firstSector, (unsigned)sector->logicalSector);
            maxSector = std::max(maxSector, (unsigned)sector->logicalSector);
            _geometry.sectorSize =
                std::max(_geometry.sectorSize, sector->data.size());
            _geometry.totalBytes += _geometry.sectorSize;
        }
    }
    _geometry.numSectors = maxSector - _geometry.firstSector + 1;
}

File: java/com/cowlark/fluxengine/data/Image.java
package com.cowlark.fluxengine.data;

import com.cowlark.fluxengine.core.Bytes;
import java.util.Collection;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.stream.Stream;

/**
 * A disk image, a collection of sectors indexed by logical location, ported
 * from lib/data/image.h.
 */
public class Image implements Iterable<Sector>
{
    private final Map<LogicalLocation, Sector> sectors = new LinkedHashMap<>();
    private Geometry geometry = new Geometry();

    public Image()
    {
    }

    public Image(Collection<Sector> sectors)
    {
        for (Sector sector : sectors)
            this.sectors.put(sector.location, sector);
        calculateSize();
    }

    public void calculateSize()
    {
        geometry = new Geometry();
        int maxSector = 0;
        for (Map.Entry<LogicalLocation, Sector> entry : sectors.entrySet())
        {
            Sector sector = entry.getValue();
            if (sector != null)
            {
                geometry.numCylinders =
                        Math.max(geometry.numCylinders, sector.location.logicalCylinder() + 1);
                geometry.numHeads = Math.max(geometry.numHeads, sector.location.logicalHead() + 1);
                geometry.firstSector =
                        Math.min(geometry.firstSector, sector.location.logicalSector());
                maxSector = Math.max(maxSector, sector.location.logicalSector());
                geometry.sectorSize = Math.max(geometry.sectorSize, sector.data.size());
                geometry.totalBytes += geometry.sectorSize;
            }
        }
        geometry.numSectors = maxSector - geometry.firstSector + 1;
    }

    public void clear()
    {
        sectors.clear();
        geometry = new Geometry();
    }

    public boolean empty()
    {
        return sectors.isEmpty();
    }

    public boolean contains(LogicalLocation location)
    {
        return sectors.containsKey(location);
    }

    public boolean contains(int cylinder, int head, int sector)
    {
        return contains(new LogicalLocation(cylinder, head, sector));
    }

    public Sector get(LogicalLocation location)
    {
        return sectors.get(location);
    }

    public Sector get(int cylinder, int head, int sector)
    {
        return get(new LogicalLocation(cylinder, head, sector));
    }

    public Sector put(LogicalLocation location)
    {
        Sector sector = new Sector(location);
        sectors.put(location, sector);
        return sector;
    }

    public Sector put(int cylinder, int head, int sector)
    {
        return put(new LogicalLocation(cylinder, head, sector));
    }

    public void erase(LogicalLocation location)
    {
        sectors.remove(location);
    }

    public void erase(int cylinder, int head, int sector)
    {
        erase(new LogicalLocation(cylinder, head, sector));
    }

    public void addMissingSectors(DiskLayout layout, boolean populated)
    {
        for (LogicalLocation location : layout.logicalSectorLocationsInFilesystemOrder)
        {
            if (!sectors.containsKey(location))
            {
                LogicalTrackLayout ltl =
                        layout.layoutByLogicalLocation.get(location.trackLocation());
                Sector sector = new Sector(location);

                if (populated)
                    sector.data = new Bytes(ltl.sectorSize);
                else
                    sector.status = Sector.Status.MISSING;

                sectors.put(location, sector);
            }
        }
        calculateSize();
    }

    public void populateSectorPhysicalLocationsFromLogicalLocations(DiskLayout diskLayout)
    {
        Image tempImage = new Image();
        for (Sector sector : this)
        {
            LogicalTrackLayout ltl =
                    diskLayout.layoutByLogicalLocation.get(sector.location.trackLocation());
            Sector newSector = tempImage.put(
                    sector.location.logicalCylinder(),
                    sector.location.logicalHead(),
                    sector.location.logicalSector());
            newSector.location = sector.location;
            newSector.status = sector.status;
            newSector.position = sector.position;
            newSector.clockNs = sector.clockNs;
            newSector.headerStartTimeNs = sector.headerStartTimeNs;
            newSector.headerEndTimeNs = sector.headerEndTimeNs;
            newSector.dataStartTimeNs = sector.dataStartTimeNs;
            newSector.dataEndTimeNs = sector.dataEndTimeNs;
            newSector.data = sector.data;
            newSector.records = sector.records;
            newSector.physicalLocation = new CylinderHead(ltl.physicalCylinder, ltl.physicalHead);
        }

        for (Sector sector : tempImage)
            sectors.put(sector.location, sector);
    }

    public Geometry getGeometry()
    {
        return geometry;
    }

    public void setGeometry(Geometry geometry)
    {
        this.geometry = geometry;
    }

    @Override
    public Iterator<Sector> iterator()
    {
        return sectors.values().iterator();
    }

    public Stream<Sector> stream()
    {
        return sectors.values().stream();
    }
}